src/main.cpp

```diff
@@ -982,32 +1044,71 @@ int main ( int argc, char** argv )
                     Server.SetDirectoryType ( AT_CUSTOM );
                 }
 
-                pApp->exec();
+                exit_code = pCoreApplication->exec();
             }
         }
     }
 
+    catch ( const CInfoExit& info )
+    {
+        // show info and exit normally
+        qInfo() << qUtf8Printable ( QString ( "%1: %2" ).arg ( strAppName, info.GetInfoMessage() ) );
+        exit_code = 0;
+    }
+
+    catch ( const CErrorExit& fatalerr )
+    {
+        // show deliberate exit message
+        qCritical() << qUtf8Printable ( QString ( "%1: %2" ).arg ( strAppName, fatalerr.GetErrorMessage() ) );
+        exit_code = fatalerr.GetExitCode();
+    }
+
     catch ( const CGenErr& generr )
     {
-        // show generic error
-#ifndef HEADLESS
-        if ( bUseGUI )
-        {
-            QMessageBox::critical ( nullptr, APP_NAME, generr.GetErrorText(), "Quit", nullptr );
-        }
-        else
+        // show unhandled generic error message
+        qCritical() << qUtf8Printable ( QString ( "%1: %2" ).arg ( strAppName, generr.GetErrorText() ) );
+        exit_code = generr.GetExitCode();
+    }
+
+    catch ( ... )
+    {
+        // show all other unhandled (standard) exceptions
+        qCritical() << qUtf8Printable ( QString ( "%1: Unknown Exception, Exiting" ).arg ( strAppName ) );
+        exit_code = -1;
+    }
+
+    if ( pServerRpc )
```

Comment:
Not needed with a smart pointer. (But even if it was, you never need to write `if (x) delete x;` - `delete NULL;` is [guaranteed to do nothing](https://isocpp.org/wiki/faq/freestore-mgmt#delete-handles-null).)

Reply:
Just a (good?) habit of me to always check for null pointers ;=)
But you are right, it's not necessary here....